src/soc/ibm/power9/istep_9_2.c

```diff
@@ -0,0 +1,311 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#include <cpu/power/istep_9.h>
+
+#include <console/console.h>
+#include <cpu/power/scom.h>
+#include <delay.h>
+#include <timer.h>
+
+#include "fsi.h"
+#include "xbus.h"
+
+struct edip_data_t {
+	uint32_t en_margin_pu;
+	uint32_t en_margin_pd;
+	uint32_t en_main;
+	uint32_t sel_pre;
+};
+
+static void compute_margin_data(uint32_t margin, struct edip_data_t *d)
+{
+	/* ATTR_IO_XBUS_TX_FFE_PRECURSOR = 6 (default, talos.xml) */
+	const uint8_t ffe_pre_coef = 6;
+
+	/* Need to convert the 8R value to a 4R equivalent */
+
+	const uint32_t val = margin >> 1;
+	const uint32_t en_pre = 18;
+
+	uint32_t val_4r = val - en_pre;
+
+	d->en_margin_pu = 32;
+	d->en_margin_pd = 32;
+	d->en_main = 0;
+	d->sel_pre = 0;
+
+	if (val_4r < 64) {
+		if (val_4r % 4 != 0) {
+			d->en_main = 2;
+			val_4r -= d->en_main;
+		}
+		d->en_margin_pd = val_4r / 2;
+		d->en_margin_pu = val_4r - d->en_margin_pd;
+	}
+
+	d->en_main += val_4r - d->en_margin_pu - d->en_margin_pd;
+	d->en_main = MIN(d->en_main, 50);
+	d->sel_pre = (val * ffe_pre_coef) / 128;
+	d->sel_pre = MIN(d->sel_pre, en_pre);
+}
+
+/* Converts a 4R decimal value to a 1R thermometer code */
+static uint32_t convert_4r(uint32_t val_4r)
+{
+	/*
+	 * 1. Add 2 for averaging since we will truncate the last 2 bits.
+	 * 2. Divide by 4 to bring back to a 1r value.
+	 * 3. Convert the decimal number to number of bits set by shifting a 0x1
+	 *    over by the amount and subtracting 1.
+	 */
+	return (0x1 << ((val_4r + 2) / 4)) - 1;
+}
+
+static uint32_t convert_4r_with_2r(uint32_t val_4r, uint8_t width)
+{
+	/* Add 1 for rounding, then shift the 4r bit off. We now have a 2r equivalent */
+	const uint32_t val_2r = (val_4r + 1) >> 1;
+
+	/* If the LSB of the 2r equivalent is on, then we need to set the 2r bit (MSB) */
+	const uint32_t on_2r = val_2r & 0x1;
+
+	/* Shift the 2r equivalent to a 1r value and convert to a thermometer code */
+	const uint32_t val_1r = (1 << (val_2r >> 0x1)) - 1;
+
+	/* Combine 1r equivalent thermometer code + the 2r MSB value */
+	return (on_2r << (width - 1)) | val_1r;
+}
+
+static void config_run_bus_group_mode(uint8_t chip, int group)
+{
+	enum {
+		P9A_XBUS_TX_IMPCAL_PVAL_PB = 0x800F140006010C3F,
+		P9A_XBUS_0_TX_IMPCAL_NVAL_PB = 0x800F0C0006010C3F,
```

Comment:
Keep the same format of register name:
```suggestion
		P9A_XBUS_TX_IMPCAL_NVAL_PB = 0x800F0C0006010C3F,
```